File: app/src/main/java/com/example/mediaplayer/chosenSong/ChosenSongService.java
package com.example.mediaplayer.chosenSong;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.os.Binder;
import android.os.IBinder;
import androidx.core.app.NotificationManagerCompat;
import com.example.mediaplayer.playlist.PlayListModel;
import com.google.android.exoplayer2.ExoPlayer;
import com.google.android.exoplayer2.ExoPlayerFactory;
import com.google.android.exoplayer2.Player;
import com.google.android.exoplayer2.source.ConcatenatingMediaSource;
import com.google.android.exoplayer2.source.MediaSource;
import com.google.android.exoplayer2.source.ProgressiveMediaSource;
import com.google.android.exoplayer2.upstream.DataSource;
import com.google.android.exoplayer2.upstream.DefaultDataSourceFactory;
import com.google.android.exoplayer2.util.Util;
import java.util.ArrayList;
import java.util.Objects;
import static com.example.mediaplayer.constants.ACTION_FOREGROUND;
import static com.example.mediaplayer.constants.CHOSEN_SONG_INDEX;
import static com.example.mediaplayer.constants.LIST_SONG;
import static com.example.mediaplayer.constants.NEXT_ACTION;
import static com.example.mediaplayer.constants.NOTIFICATION_ID;
import static com.example.mediaplayer.constants.PAUSE_ACTION;
import static com.example.mediaplayer.constants.PLAY_ACTION;
import static com.example.mediaplayer.constants.PREVIOUS_ACTION;

public class ChosenSongService extends Service {

    public ExoPlayer getPlayer() {
        return player;
    }

    private ExoPlayer player;
    private boolean mPlayWhenReady = true;
    private int currentWindowIndex = 0;
    private Context application;

    // indicates how to behave if the service is killed.
    int mStartMode = START_STICKY;

    // interface for clients that bind.
    IBinder mBinder = new SongBinder();
    //responsible for creating media player notification;
    NotificationController notificationController;
    //responsible for updating the notification
    NotificationManagerCompat mNotifymanager;
    //playList of songs
    ArrayList<PlayListModel> songList;

    @Override
    public void onCreate() {
        // The service is being created.
        player = ExoPlayerFactory.newSimpleInstance(getApplicationContext());
        mNotifymanager = NotificationManagerCompat.from(ChosenSongService.this);
        application = getApplicationContext();

    }

    //creating concatenating media source for media player to play
    private MediaSource buildMediaSource(ArrayList<PlayListModel> audioUris) {
        // Produces DataSource instances through which media data is loaded.
        DataSource.Factory dataSourceFactory = new DefaultDataSourceFactory(application,
                Util.getUserAgent(application, "MediaPlayer"));
        ConcatenatingMediaSource concatenatingMediaSource = new ConcatenatingMediaSource();
        if (audioUris == null) {
            return null;
        } else {
            for (PlayListModel item : audioUris) {
                concatenatingMediaSource.addMediaSource(new ProgressiveMediaSource.Factory(dataSourceFactory)
                        .createMediaSource(item.getAudioUri()));
            }
        }
        return concatenatingMediaSource;
    }

    public void setUpPlayer(ArrayList<PlayListModel> audioUris, int chosenSongIndex) {

        //to control to player the audio or video right now or wait user to play the audio himself
        player.setPlayWhenReady(true);
        MediaSource mediaSource = buildMediaSource(audioUris);
        player.prepare(mediaSource);
        //to control the starter location of audio
        player.seekTo(chosenSongIndex, 0);


    }
    private void handlePlayerEvent() {
        player.addListener(new Player.EventListener() {
            @Override
            public void onPlayerStateChanged(boolean playWhenReady, int playbackState) {
                if (isPlayerStopped(playWhenReady)) {
                    mPlayWhenReady = false;
                    mNotifymanager.notify(NOTIFICATION_ID, notificationController.buildNotification(mPlayWhenReady, currentWindowIndex));

                } else if (isPlayerPlaying(playWhenReady)) {
                    mPlayWhenReady = true;
                    mNotifymanager.notify(NOTIFICATION_ID, notificationController.buildNotification(mPlayWhenReady, currentWindowIndex));


                } else if (isPlayerNext()) {
                    currentWindowIndex = player.getCurrentWindowIndex();
                    mNotifymanager.notify(NOTIFICATION_ID, notificationController.buildNotification(player.getPlayWhenReady(), currentWindowIndex));


                } else if (isPlayerPrevious()) {
                    currentWindowIndex = player.getCurrentWindowIndex();
                    mNotifymanager.notify(NOTIFICATION_ID, notificationController.buildNotification(player.getPlayWhenReady(), currentWindowIndex));

                }


            }

        });
    }

    private boolean isPlayerStopped(boolean playWhenReady) {
        return playWhenReady != mPlayWhenReady && !playWhenReady;
    }

    private boolean isPlayerPlaying(boolean playWhenReady) {
        return playWhenReady != mPlayWhenReady && playWhenReady;
    }

    private boolean isPlayerNext() {
        return player.getCurrentWindowIndex() != currentWindowIndex && player.getCurrentWindowIndex() < currentWindowIndex;

    }

    private boolean isPlayerPrevious() {
        return player.getCurrentWindowIndex() != currentWindowIndex && player.getCurrentWindowIndex() > currentWindowIndex;

    }


    class SongBinder extends Binder {
        ChosenSongService getService() {
            return ChosenSongService.this;
        }

    }

    private void handleIntent(Intent intent) {
        if (intent != null) {
            switch (Objects.requireNonNull(intent.getAction())) {
                case ACTION_FOREGROUND:
                    //getting the playlist of song from intent
                    songList = intent.getParcelableArrayListExtra(LIST_SONG);
                    //getting the current playing song index
                    currentWindowIndex = intent.getIntExtra(CHOSEN_SONG_INDEX, 0);
                    notificationController = new NotificationController(songList, this);
                    handlePlayerEvent();
                    startForeground(NOTIFICATION_ID, notificationController.buildNotification(mPlayWhenReady, currentWindowIndex));
                    break;
                case PAUSE_ACTION:
                    player.setPlayWhenReady(false);
                    break;
                case PLAY_ACTION:
                    player.setPlayWhenReady(true);
                    break;
                case PREVIOUS_ACTION:
                    player.previous();
                    break;
                case NEXT_ACTION:
                    player.next();
                    break;
            }

        }

    }

    @Override
    public int onStartCommand(Intent intent,
                              int flags, int startId) {
        // The service is starting, due to a call to startService().
        handleIntent(intent);
        return mStartMode;
    }

    @Override
    public IBinder onBind(Intent intent) {
        return mBinder;
    }

    @Override
    public void onDestroy() {
        // The service is no longer used and is being destroyed
    }

}
//playing an audio media on device ,app is using foreground service with banded service to gracefully handle any configuration changes and ipc.this app is using ExoPlayer library

File: app/src/main/java/com/example/mediaplayer/forgroundService/ChosenSongService.kt
package com.example.mediaplayer.forgroundService

import android.app.Service
import android.content.Context
import android.content.Intent
import android.os.Binder
import android.os.IBinder
import androidx.core.app.NotificationManagerCompat
import com.example.mediaplayer.*
import com.example.mediaplayer.model.PlayListModel
import com.google.android.exoplayer2.C
import com.google.android.exoplayer2.ExoPlayerFactory
import com.google.android.exoplayer2.Player
import com.google.android.exoplayer2.SimpleExoPlayer
import com.google.android.exoplayer2.audio.AudioAttributes
import com.google.android.exoplayer2.source.ConcatenatingMediaSource
import com.google.android.exoplayer2.source.MediaSource
import com.google.android.exoplayer2.source.ProgressiveMediaSource
import com.google.android.exoplayer2.upstream.DefaultDataSourceFactory
import com.google.android.exoplayer2.util.Util
import java.util.*


class ChosenSongService : Service() {

    lateinit var player: SimpleExoPlayer
        private set
    private var mPlayWhenReady = true
    private var currentWindowIndex = 0
    private var application: Context? = null

    // indicates how to behave if the service is killed.
    private var mStartMode = START_STICKY

    // interface for clients that bind.
    private var mBinder: IBinder = SongBinder()
    //responsible for creating media player notification;
    lateinit var notificationController: NotificationController
    //responsible for updating the notification
    lateinit var mNotifymanager: NotificationManagerCompat

    private val isPlayerNext: Boolean
        get() = player.currentWindowIndex != currentWindowIndex && player.currentWindowIndex < currentWindowIndex

    private val isPlayerPrevious: Boolean
        get() = player.currentWindowIndex != currentWindowIndex && player.currentWindowIndex > currentWindowIndex

    override fun onCreate() {
        // The service is being created.
        player = ExoPlayerFactory.newSimpleInstance(applicationContext)
        mNotifymanager = NotificationManagerCompat.from(this@ChosenSongService)
        application = applicationContext

    }

    //creating concatenating media source for media player to play
    private fun buildMediaSource(audioUris: ArrayList<PlayListModel>?): MediaSource? {
        // Produces DataSource instances through which media data is loaded.
        val dataSourceFactory = DefaultDataSourceFactory(application!!,
                Util.getUserAgent(application, "MediaPlayer"))
        val concatenatingMediaSource = ConcatenatingMediaSource()
        if (audioUris == null) {
            return null
        } else {
            for (item in audioUris) {
                concatenatingMediaSource.addMediaSource(ProgressiveMediaSource.Factory(dataSourceFactory)
                        .createMediaSource(item.audioUri))
            }
        }
        return concatenatingMediaSource
    }

    fun setUpPlayer(audioUris: ArrayList<PlayListModel>, chosenSongIndex: Int) {
        //to control to player the audio or video right now or wait user to play the audio himself
        player.playWhenReady = true
        val mediaSource = buildMediaSource(audioUris)
        player.prepare(mediaSource)
        //to control the starter location of audio
        player.seekTo(chosenSongIndex, 0)
        //to handle audio focus changes
        player.setAudioAttributes(AudioAttributes.Builder()
                .setUsage(C.USAGE_MEDIA)
                .setContentType(C.CONTENT_TYPE_MUSIC)
                .build(), true)


    }

    private fun handlePlayerEvent() {
        player.addListener(object : Player.EventListener {
            override fun onPlayerStateChanged(playWhenReady: Boolean, playbackState: Int) {
                when {
                    isPlayerStopped(playWhenReady) -> {
                        mPlayWhenReady = false
                        mNotifymanager.notify(NOTIFICATION_ID, notificationController.buildNotification(mPlayWhenReady, currentWindowIndex))

                    }
                    isPlayerPlaying(playWhenReady) -> {
                        mPlayWhenReady = true
                        mNotifymanager.notify(NOTIFICATION_ID, notificationController.buildNotification(mPlayWhenReady, currentWindowIndex))


                    }
                    isPlayerNext -> {
                        currentWindowIndex = player.currentWindowIndex
                        mNotifymanager.notify(NOTIFICATION_ID, notificationController.buildNotification(player.playWhenReady, currentWindowIndex))


                    }
                    isPlayerPrevious -> {
                        currentWindowIndex = player.currentWindowIndex
                        mNotifymanager.notify(NOTIFICATION_ID, notificationController.buildNotification(player.playWhenReady, currentWindowIndex))

                    }
                }


            }

        })
    }

    private fun isPlayerStopped(playWhenReady: Boolean): Boolean {
        return playWhenReady != mPlayWhenReady && !playWhenReady
    }

    private fun isPlayerPlaying(playWhenReady: Boolean): Boolean {
        return playWhenReady != mPlayWhenReady && playWhenReady
    }


    internal inner class SongBinder : Binder() {
        val service: ChosenSongService
            get() = this@ChosenSongService

    }

    private fun handleIntent(intent: Intent?) {
        if (intent != null) {
            when (Objects.requireNonNull<String>(intent.action)) {
                ACTION_FOREGROUND -> {
                    //getting the playlist of song from intent
                    //playList of songs
                    val songList: ArrayList<PlayListModel>? = intent.getParcelableArrayListExtra(LIST_SONG)
                    //getting the current playing song index
                    currentWindowIndex = intent.getIntExtra(CHOSEN_SONG_INDEX, 0)
                    notificationController = NotificationController(songList, this)
                    handlePlayerEvent()
                    startForeground(NOTIFICATION_ID, notificationController.buildNotification(mPlayWhenReady, currentWindowIndex))
                }
                PAUSE_ACTION -> player.playWhenReady = false
                PLAY_ACTION -> player.playWhenReady = true
                PREVIOUS_ACTION -> player.previous()
                NEXT_ACTION -> player.next()
            }

        }

    }

    override fun onStartCommand(intent: Intent?,
                                flags: Int, startId: Int): Int {
        // The service is starting, due to a call to startService().
        handleIntent(intent)
        return mStartMode
    }

    override fun onBind(intent: Intent): IBinder? {
        return mBinder
    }


}
